packages/deployments/contracts/contracts/messaging/libraries/Queue.sol

```diff
@@ -106,8 +109,15 @@ library QueueLib {
     uint256 index; // Cursor for index in the batch of `items`.
     // NOTE: `first <= last` rephrased here to `!(first > last)` as it's a cheaper condition.
     while (!(first > last)) {
-      items[index] = queue.data[first];
+      bytes32 item = queue.data[first];
+      // Check to see if the item has been removed before appending it to the array.
+      if (!queue.removed[item]) {
+        items[index] = item;
```

Comment:
This leaves empty spaces in `items[]`. (note the 0 values could be tricky)
Because `index` is increased even when `removed[]` is `true`

Reply:
hmm. Open to any suggestions here, cc @csanuragjain since your comment seems to be saying the same thing.

I definitely don't want to leave gaps, but the idea of copying over into another array in memory does not sound cost effective

Reply:
Alternatively we could just *not* increment index if `removed[]` is `true` and instead return an array with some potential subset of 0-values at the end of it, and document that in the natspec here. We could then handle this particular arrangement wherever the array ends up being consumed.

Otherwise, there might be a hacky way to truncate the array in assembly of its 0-values...

Reply:
Actually, I might just elect for simpler solution of copying over to a second return-value array IFF there was a removed item. Basically, track whether an item was removed in this loop, then recopy accordingly. That way if we're going gas-inefficient route, at least it will only be inefficient in edge case